[AOTI] Support scalar to tensor in the ABI-compatible mode

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118024



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

Differential Revision: [D53019485](https://our.internmc.facebook.com/intern/diff/D53019485)

desertfire committed:
[AOTI] Support scalar to vector in the ABI-compatible mode

[ghstack-poisoned]

desertfire committed:
Update on "[AOTI] Support scalar to tensor in the ABI-compatible mode"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ezyang commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -135,6 +141,31 @@ AOTI_TORCH_ITEM_IMPL(int32, int32_t)
+// AOTI_TORCH_VALUE_TO_TENSOR_IMPL(uint32, uint32_t, UInt32)
+// error: conversion from ‘uint64_t’ {aka ‘long unsigned int’} to
+// ‘torch::detail::TensorDataContainer’ is ambiguous
+// AOTI_TORCH_VALUE_TO_TENSOR_IMPL(uint64, uint64_t, UInt64)
```
Hmm... Well you could always use scalar_to_tensor instead, that should work...

desertfire committed:
Update on "[AOTI] Support scalar to tensor in the ABI-compatible mode"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

desertfire commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -56,6 +57,11 @@ c10::optional<T> pointer_to_optional(U* ptr) {
 
+AtenTensorHandle new_tensor_handle(at::Tensor&& tensor) {
+  at::Tensor* new_tensor = new at::Tensor(std::move(tensor));
+  return tensor_pointer_to_tensor_handle(new_tensor);
```
It should, but somehow everybody has been just doing copy-pasting. I will post another refactor PR after this.

desertfire commented on torch/_inductor/ir.py:
```diff
@@ -4417,6 +4426,8 @@ def __init__(self, sym, data):
             self.sym = sym.args[0]
             self.is_bool = True
 
+        V.graph.symbol_to_dtype[self.sym] = data.get_dtype()
```
If we return another data struct, say a data class with symbol and dtype, instead of symbol here, https://github.com/pytorch/pytorch/blob/e309d6fa1c4c3c7fe3f192e40f4aec4b4609c0f5/torch/_inductor/lowering.py#L2558. We need to take care of all the following phases that may expect a symbol type. I am not sure if that's worth to do just for the purpose of this PR.

ezyang replied:
No, I still think this is very suspicious.

Looking over the modifications you made, the change here is solely so that ShapeAsConstantBuffer can be code generated.  But here, there is a strange divergence between the Python in C++. In Python, the shape is code generated as a plain int; however, in C++, the shape is wrapped into a Tensor. This seems... wrong! Now that we are supporting raw integers flowing through AOTInductor, the shape should be represented as is (as a `int64_t`) when flowing through AOTInductor.

desertfire committed:
Update on "[AOTI] Support scalar to tensor in the ABI-compatible mode"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D53019485).

desertfire committed:
Update on "[AOTI] Support scalar to tensor in the ABI-compatible mode"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D53019485](https://our.internmc.facebook.com/intern/diff/D53019485)

[ghstack-poisoned]

desertfire committed:
Update on "[AOTI] Support scalar to tensor in the ABI-compatible mode"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D53019485](https://our.internmc.facebook.com/intern/diff/D53019485)

[ghstack-poisoned]

desertfire committed:
Update on "[AOTI] Support scalar to tensor in the ABI-compatible mode"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D53019485](https://our.internmc.facebook.com/intern/diff/D53019485)

[ghstack-poisoned]

ezyang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1897,6 +1915,15 @@ def use_thread_local_cached_output_tensor(idx, output):
                 if config.aot_inductor.abi_compatible:
+                    output_buffer = V.graph.graph_outputs[idx]
+                    if isinstance(output_buffer, ir.ShapeAsConstantBuffer):
+                        # Need to wrap scalar into tensor as the main function returns a vector of tensors
```
HMMMM are you ever considering to change this ABI? I can imagine that returning shapes from AOTInductor models is not that common today and so maybe papering over the ABI problem this way is fine, but when we support dynamic shapes AOTInductor models, I *really* don't want input ints to be passed into the model as tensors, it's soooo unnecessary.

cc @oulgen @bdhirsh

desertfire replied:
cc @chenyang78  who may look into this later.

ezyang reviewed (approved):
Thanks, looks good!

desertfire committed:
Update on "[AOTI] Support scalar to tensor in the ABI-compatible mode"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D53019485](https://our.internmc.facebook.com/intern/diff/D53019485)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D53019485).
